pex/pep_427.py

```diff
@@ -0,0 +1,310 @@
+    compile=False,  # type: bool
+    requested=True,  # type: bool
+):
+    # type: (...) -> MetadataFiles
+
+    # See: https://packaging.python.org/en/latest/specifications/binary-distribution-format/#installing-a-wheel-distribution-1-0-py32-none-any-whl
+    metadata_files = load_metadata(wheel_path, restrict_types_to=(MetadataType.DIST_INFO,))
+    if not metadata_files:
+        raise ValueError("Could not find any metadata in {wheel}.".format(wheel=wheel_path))
+
+    wheel_metadata_path = metadata_files.metadata_file_rel_path("WHEEL")
+    wheel_metadata = metadata_files.read("WHEEL")
+    if not wheel_metadata_path or not wheel_metadata:
+        raise ValueError("Could not find WHEEL metadata in {wheel}.".format(wheel=wheel_path))
+    wheel_metadata_dir = os.path.dirname(wheel_metadata_path)
+    if not wheel_metadata_dir.endswith(".dist-info"):
+        raise ValueError(
+            "Expected WHEEL metadata for {wheel} to be housed in a .dist-info directory, but was "
+            "found at {wheel_metadata_path}.".format(
+                wheel=wheel_path, wheel_metadata_path=wheel_metadata_path
+            )
+        )
+
+    purelib = "true" == parse_message(wheel_metadata).get("Root-Is-Purelib")
+    dest = install_paths.purelib if purelib else install_paths.platlib
+
+    record_relpath = os.path.join(wheel_metadata_dir, "RECORD")
+    record_abspath = os.path.join(dest, record_relpath)
+
+    data_rel_path = re.sub(r"\.dist-info$", ".data", wheel_metadata_dir)
+    data_path = os.path.join(dest, data_rel_path)
+
+    installed_files = []  # type: List[InstalledFile]
+
+    def record_files(
+        root_dir,  # type: Text
+        names,  # type: Iterable[Text]
+    ):
+        # type: (...) -> None
+        for name in sorted(names):
+            if is_pyc_file(name):
+                # These files are both optional to RECORD and should never be present in wheels
+                # anyway per the spec.
+                continue
+            file_abspath = os.path.join(root_dir, name)
+            if record_relpath == name:
+                # We'll generate a new RECORD below.
+                os.unlink(file_abspath)
+                continue
+            installed_files.append(
+                InstalledWheel.create_installed_file(path=file_abspath, dest_dir=dest)
+            )
+
+    with open_zip(wheel_path) as zf:
+        zf.extractall(dest)
+        # TODO(John Sirois): Consider verifying signatures.
+        # N.B.: Pip does not and its also not clear what good this does. A zip can be easily poked
+        # on a per-entry basis allowing forging a RECORD entry and its associated file. Only an
+        # outer fingerprint of the whole wheel really solves this sort of tampering.
+        record_files(
+            root_dir=dest,
+            names=[
+                name
+                for name in zf.namelist()
+                if not name.endswith("/") and data_rel_path != commonprefix((data_rel_path, name))
```

Comment:
Yeah, definitely not handling Windows here. As far as `name.startswith(data_rel_path)`, this is ~always possibly incorrect and so I ~never allow myself that test. Say `data_rel_path = "a/path"` and `name = "a/path-different/foo".

Reply:
Sorry, the second paragraph in my comment wasn't very clear, but I was trying to point out that since you use `commonprefix` and not `commonpath` this code is already susceptible to exactly that problem.

Reply:
```
>>> import os
>>> data_rel_path = "a/path"
>>> name = "a/path-different/foo"
>>> os.path.commonprefix((data_rel_path, name))
'a/path'
```

Reply:
Aha, bad import. I went through the trouble of porting `commonpath` for Python 2 in `pex.compatibility` a while back for my stated reason! Thanks.

Reply:
Fixed.